Cherry pick the `src/modules/land_detector/` directory work from PR #9756

mcsauder opened the pull request:
**Describe problem solved by the proposed pull request**
This PR captures the remaining `src/modules/land_detector/` directory changes from PR #9756 to simplify that PR.

The changes in the `land_detector` directory are fairly straightforward:
 - Break out common features of the child class functionality and move to the `LandDetector::initialize_topics()` and `LandDetector::update_topics()` parent class methods.
 - Standardize class member variable naming, uniform initialization, and whitespace formatting across all class implementations in the land_detector module. 

This PR primarily aims to reduce the scope of PR #9756.

**Describe possible alternatives**
Keeping the current implementations is certainly a viable option; please let me know if that is the preference!

**Additional context**
See PR #9756.

Much of the original work proposed in this PR has already been carried out by #12209, #12677, #12702, #13079, and #13077.

Thanks!

-Mark

mcsauder commented on src/modules/land_detector/MulticopterLandDetector.cpp:
```diff
@@ -131,70 +134,76 @@ void MulticopterLandDetector::_update_params()
-		// Widen acceptance thresholds for landed state right after arming
-		// so that motor spool-up and other effects do not trigger false negatives.
-		verticalMovement = fabsf(_vehicleLocalPosition.vz) > _params.maxClimbRate  * 2.5f;
+	if (_vehicle_local_position.v_z_valid) {
```
This added check is really the only material modification contained in this PR.

mcsauder commented:
@dagar , @LorenzMeier , please feel free to drop this PR if the work is not being pursued, I was only pushing it because of the tag.

mcsauder commented:
@dagar, the work in this PR will consolidate the meaningful work in #9756. I think this PR is in good shape now that 1.9.0 is mostly there.  If anyone else could review this (your) work I think it would be easy to get #9756 across the finish line.  Thanks for creating these improvements!

mcsauder commented:
@dagar , this is your work, is there any chance the @PX4/testflights could check this out to help progress #9756?

mcsauder commented:
@dagar, finally, I can help with flight testing.
This PR:
https://review.px4.io/plot_app?log=72c9c408-e52b-4b1c-b8c0-a4e21f2fb2b8
https://review.px4.io/plot_app?log=60098390-cb97-4338-8aea-0debd2c0c358
https://review.px4.io/plot_app?log=dfe950cd-854d-4c1d-be41-af1ec35824d4

PX4 master branch on the same drone:
https://review.px4.io/plot_app?log=e59fc2e3-702a-4e83-b1c1-66e08b2fa9d6
https://review.px4.io/plot_app?log=225d2309-4dee-48de-a820-e4297dc47ea2
https://review.px4.io/plot_app?log=5cadb268-8894-42ac-8bf4-c4da2a8b1734

The multicopter land detector seemed to behave the same for this PR and PX4 master with my airframe, neither reliably detects.  I attribute the inability to detect the landing in this case on somewhat high levels of vibration, (something I can look into on my rig), but again, no perceptible differences between this PR and master.

Please let me know if you need more data collected against this PR.

-Mark

mcsauder commented:
@dagar, is this PR of interest to you?  Your merge into master today created a substantial tangle with this PR.  I'm willing to spend the time untangling if we can advance this PR quickly when I get it done.  Let me know your thoughts.

mcsauder commented:
Initial flight test on the current rebase is satisfactory, results like the logs above.  I will post additional logs as soon as possible.

mcsauder commented:
Flight testing on a 250 quad gives expected results comparable to PX4 master:
https://review.px4.io/plot_app?log=b4f7aca7-b234-4e0e-8ab2-0eae5c1abb13
https://review.px4.io/plot_app?log=8e59ba19-a44c-4149-b493-1c22adb9f619
https://review.px4.io/plot_app?log=5f073c29-0da9-4838-b040-0a3bc6d95630

mcsauder commented:
@dagar, this PR contains a lot of variable renaming... Let me explain.  The most recent LandDetector merge contains a large number of variables that are CamelCase, snake_case, and camelSnake_case... The reason this is very bad is because of tab autocomplete with modern IDE's... mixing cases like this allows patterns to go unrecognized and variables to be duplicated with only a capitalization difference.... This is actually very dangerous mostly because of modern IDE's, see AV Rule #48: http://www.stroustrup.com/JSF-AV-rules.pdf.

That's why this PR was worth the time to fix, worth the time to flight test, and worth the time to review!

mcsauder commented:
@dagar and @bkueng , this PR now implements `ModuleParams` inheritance and the `DEFINE_PARAMETERS_CUSTOM_PARENT()` macro throughout the module.  I will post multicopter flight logs shortly.  Please see PR #12209.

mcsauder commented:
Now that this PR is down to mostly deletions, I am hopeful that it can be finished off soon!  Please let me know what else you would like to see in this PR.

Thanks!

MaEtUgR commented on src/modules/land_detector/LandDetector.cpp:
```diff
@@ -164,6 +156,11 @@ void LandDetector::_check_params(const bool force)
 	}
 }
 
+void LandDetector::_update_params()
```
How about just declaring a default parameter for `_update_params()`?
I'll add the suggestion at the declation line.

MaEtUgR replied:
I missread it now I see that it's already overloaded... I think others might struggle as well since it's quite confusing.

mcsauder replied:
Thanks @MaEtUgR , I agree and like your suggestion.  I've pushed up a commit to address.  I need to re-flight test and will get that done in a few hours then post logs.

MaEtUgR replied:
Sounds good 👍

MaEtUgR commented on src/modules/land_detector/LandDetector.h:
```diff
@@ -153,14 +159,27 @@ class LandDetector : public ModuleBase<LandDetector>, ModuleParams, px4::Schedul
+	 * Updates parameters if changes have occurred or if forced update.
+	 * @param force Force a parameter update.
+	 */
+	void _update_params(const bool force);
```
```suggestion
	void _update_params(const bool force = false);
```
Instead of overloading with no parameter?

MaEtUgR commented:
> The main functional implication of any importance in this PR is in the MulticopterLandDetector::_get_ground_contact_state() method logic.

I can't find this. It was probably removed right?

mcsauder commented:
I need to re-flight test this PR with master branch merge and rebase.  I will post flight logs asap.

mcsauder commented:
To help move this PR along I've added a commit to revert changes to the `FixedwingLandDetector::_get_max_altitude()` method that I cannot flight test at present.  This change remains an active PR in #12343.

mcsauder commented:
I've squashed all of the commits in this PR and corrected duplicate (inherited) code in the MulticopterLandDetector that I had missed pruning.

@PX4/testflights , now that it's passing CI, (only if you have time and its not too inconvenient), would you be able to try this PR on a fixed wing?  I'll get multicopter flight logs posted here as soon as I have my hardware flight ready again; should only be a day or two.

dannyfpv commented:
Tested on pixhawk1 v3 phantom wing
Modes Tested:
Takeoff: no issues
rtl Mode:. no issues
mission mode: no issues
Altitude Mode:. no issues
Stabilized Mode:. no issues
Landing:  no issues

log:  
 https://review.px4.io/plot_app?log=3672b9da-c0b9-4b53-bc70-f350517c0823
https://review.px4.io/plot_app?log=812f608d-0a99-476f-a44a-96685b82e953

Junkim3DR commented:
### Tested on Pixhawk 4mini v5:

**Modes Tested**
- Position Mode: Good.
- Altitude Mode: Good.
- Stabilized Mode: Good.
- Mission Plan Mode (Automated): Good.
- RTL (Return To Land): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint then trigger RTL and see landing behavior.

**Notes**
No issues noted, good flight in general.

**Log:**
https://review.px4.io/plot_app?log=818a58e3-4679-47ea-a2bf-f1b1d505c017

dannyfpv commented:
Tested on pixhawk4 v5 f450
Modes Tested:
Takeoff: no issues
rtl Mode:. no issues
mission mode: no issues
Altitude Mode:. no issues
Stabilized Mode:. no issues
Landing: no issues

log:
https://review.px4.io/plot_app?log=3de37f3e-4fb2-4287-8bf6-456b6331de1d

Tony3dr commented:
@mcsauder let us know if you need further testing.

mcsauder commented:
Thank you so much @Tony3dr , @dannyfpv ,and @Junkim3DR !!  I appreciate it very much.

My flight testing also showed good results, correct land detection 5 of 5 on a 250 quad with pixhawk 4 mini:

 - https://review.px4.io/plot_app?log=1f3ddca3-ff70-4fa3-a608-aa9e7f26bfaf
 - https://review.px4.io/plot_app?log=52a7fc19-65c8-418f-a6db-3924660ba173
 - https://review.px4.io/plot_app?log=4df832c9-ad11-4de3-ba84-20a3f2680d59
 - https://review.px4.io/plot_app?log=ef66132b-75e8-4dd6-80a2-051b10e5f8b6
 - https://review.px4.io/plot_app?log=04d58db3-c5b0-438f-816b-a83e996ca658

@MaEtUgR , this PR should be good to go!

bkueng commented on src/modules/land_detector/LandDetector.cpp:
```diff
@@ -71,15 +63,16 @@ LandDetector::~LandDetector()
 void LandDetector::start()
 {
-	_check_params(true);
+	_force_param_update = true;
```
Why do you use a class attribute here? It adds another state, whereas passing an argument would be simpler.

mcsauder replied:
Hi @bkueng , I changed it to meet your suggestion and flight testing is good: https://review.px4.io/plot_app?log=a7b84e5b-14dc-4324-a324-986aac6873dc

Initially I had overloaded the method to allow inheritance from the base class while preserving the original behavior.  This was changed after feedback above to use a class variable.  As it stands now, the MulticopterLandDetector class doesn't use the argument, but I don't see any issues with that.

Thanks for the feedback!

mcsauder commented:
Hi @MaEtUgR , the CI tools failed after the lastest master branch merge, so I rebased, squashed commits, and updated the commit comments to better reflect the changes that have occurred.  No changes to the code, only the commit comments and squash.

Hopefully CI is happy this round!

mcsauder commented:
Thanks @MaEtUgR ! I am excited to see this one through to the end!

I appreciate everyone's reviews and all of your feedback!

bkueng commented on src/modules/land_detector/MulticopterLandDetector.cpp:
```diff
@@ -86,17 +86,19 @@ MulticopterLandDetector::MulticopterLandDetector()
+
 	_freefall_hysteresis.set_hysteresis_time_from(false, (hrt_abstime)(1e6f * _param_lndmc_ffall_ttri.get()));
 
 	param_get(_paramHandle.minThrottle, &_params.minThrottle);
```
@mcsauder this is not correct, it re-loads the params every cycle now, as there is no check for `_parameter_update_sub` here. You can move the `_parameter_update_sub.updated()` check out of  `LandDetector::_update_params` around the `_update_params();` in `LandDetector::Run()` (and get rid of the `force` argument).
(Alternatively you can rename this method here to `void updateParams() override;`, and then make `LandDetector::_update_params` non-virtual)
There also needs to be a call to `updateParams()` in `LandDetector::_update_params`.
